### PR TITLE
[#9382] Add support for Msq type questions

### PR DESCRIPTION
**Pull request by RonakLakhotia**
Part of #9382

### CODE REVIEW

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java [hunk 1]
```diff
@@ -46,8 +46,8 @@ public FeedbackMsqQuestionDetails() {
         this.msqChoices = new ArrayList<>();
         this.otherEnabled = false;
         this.generateOptionsFor = FeedbackParticipantType.NONE;
-        this.maxSelectableChoices = Integer.MIN_VALUE;
-        this.minSelectableChoices = Integer.MIN_VALUE;
+        this.maxSelectableChoices = 0;
+        this.minSelectableChoices = 0;
```
Comment (RonakLakhotia) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: @wkurniawan07 I have a doubt. 

I used 0 instead of Integer.MIN_VALUE to check if min/max choices are enabled. The reason is `Number.MIN_VALUE` used in the default structure also decimal numbers. So if the user does not select min/max checkbox, the backend will throw a `Server cannot process your request` error because of the decimal value.

I also considered using `Number.MIN_SAFE_INTEGER` however this calculates the min integer up to 53 bits, while 
` Integer.MIN_VALUE`  calculates the min integer up to 31 bits.

File: src/web/services/feedback-questions.service.ts [hunk 2]
```diff
@@ -308,6 +315,30 @@ export class FeedbackQuestionsService {
...
+            FeedbackVisibilityType.GIVER_TEAM_MEMBERS],
+          showGiverNameTo: [FeedbackVisibilityType.INSTRUCTORS],
+          showRecipientNameTo: [FeedbackVisibilityType.INSTRUCTORS, FeedbackVisibilityType.RECIPIENT],
+        };
+      case FeedbackQuestionType.MSQ:
+
+        const msqQuestionDetails: FeedbackMsqQuestionDetails = DEFAULT_MSQ_QUESTION_DETAILS();
+        msqQuestionDetails.msqChoices = [' ', ' '];
+        msqQuestionDetails.minSelectableChoices = 0;
+        msqQuestionDetails.maxSelectableChoices = 0;
```
Comment (RonakLakhotia) [inline src/web/services/feedback-questions.service.ts]: changed this from `Number.MIN_VALUE` to 0

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java (hunk 1 above)
Comment (wkurniawan07) [reply src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: `0` is a legitimate option value for min. selectable choices. Why not `-1`?

Comment (RonakLakhotia) [reply src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: I thought the minimum selectable choices must be greater than `0`.

```java
        if (isMinSelectableChoicesEnabled && minSelectableChoices < 1) {
            errors.add(Const.FeedbackQuestion.MSQ_ERROR_MIN_FOR_MIN_SELECTABLE_CHOICES);
        }
```

This is the existing backend validation for `FeedbackMsqQuestionDetails`

Comment (wkurniawan07) [reply src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: @damithc has it always been the case that MSQ questions must have min. choice of 1?

This question:
```
Select programming languages that you know:
- Java
- C++
- ...
```
Looks like it should accept zero choice as an acceptable answer.

Comment (RonakLakhotia) [reply src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]:
```java
        if (isMinSelectableChoicesEnabled && minSelectableChoices < 1) {
            errors.add(Const.FeedbackQuestion.MSQ_ERROR_MIN_FOR_MIN_SELECTABLE_CHOICES);
        }
```
The min selectable choices can be 0 but if the checkbox is enabled, the min choices must be one as indicated by the validation

Comment (damithc) [reply src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: The rationale is that specifying 0 as the minimum is the same as specifying there is no minimum number of options; in that case, the user can simply not enable the constraint at all.

**Comment by RonakLakhotia** [comment]
@wkurniawan07 read for review

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java (hunk 1 above)
Comment (wkurniawan07) [reply src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: @RonakLakhotia that still does not change the fact that `0` is a value that can be legitimately set by user, not a value that should be specially treated by the system.

File: src/web/app/components/question-types/question-edit-details-form/msq-question-edit-details-form.component.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  /**
+   * Displays minSelectableOption value.
+   */
+  get displayValueForMinSelectableOption(): number {
+    if (this.model.minSelectableChoices === 0) {
+      return 1;
+    }
+    return this.model.minSelectableChoices;
```
Comment (wkurniawan07) [inline src/web/app/components/question-types/question-edit-details-form/msq-question-edit-details-form.component.ts]: You can shorten this to:
```ts
return this.model.minSelectableChoices || 1;
```

File: src/web/app/components/question-types/question-edit-details-form/msq-question-edit-details-form.component.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  /**
+   * Displays maxSelectableOption value.
+   */
+  get displayValueForMaxSelectableOption(): number {
+    if (this.model.maxSelectableChoices === 0) {
+      return 2;
+    }
+    return this.model.maxSelectableChoices;
```
Comment (wkurniawan07) [inline src/web/app/components/question-types/question-edit-details-form/msq-question-edit-details-form.component.ts]: You can shorten this to:
```ts
return this.model.maxSelectableChoices || 2;
```

File: src/web/app/components/question-types/question-edit-answer-form/msq-question-edit-answer-form.component.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+@Component({
+  selector: 'tm-msq-question-edit-answer-form',
+  templateUrl: './msq-question-edit-answer-form.component.html',
+  styleUrls: ['./msq-question-edit-answer-form.component.scss'],
+})
+export class MsqQuestionEditAnswerFormComponent
+    extends QuestionEditAnswerFormComponent<FeedbackMsqQuestionDetails, FeedbackMsqResponseDetails> implements OnInit {
+
+  isMsqOptionSelected: boolean[] = Array(this.questionDetails.msqChoices.length).fill(false);
+  readonly NONE_OF_THE_ABOVE: string = 'None of the above';
```
Comment (wkurniawan07) [inline src/web/app/components/question-types/question-edit-answer-form/msq-question-edit-answer-form.component.ts]: This seems like it's only used inside this class and not in the HTML template. Let's make this `private`, or even better, move this to outside the class definition (i.e. make it `const NONE_OF_THE_ABOVE = ...`).

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqResponseDetails.java [hunk 6]
```diff
@@ -85,7 +85,59 @@ public String getAnswerCsv(FeedbackQuestionDetails questionDetails) {
...
-        return new ArrayList<>();
+        List<String> errors = new ArrayList<>();
+        List<String> msqChoices = ((FeedbackMsqQuestionDetails) correspondingQuestion
+                .getQuestionDetails()).getMsqChoices();
+        int maxSelectableChoices = ((FeedbackMsqQuestionDetails) correspondingQuestion
+                .getQuestionDetails()).getMaxSelectableChoices();
+        int minSelectableChoices = ((FeedbackMsqQuestionDetails) correspondingQuestion
+                .getQuestionDetails()).getMinSelectableChoices();
+        boolean isOtherEnabled = ((FeedbackMsqQuestionDetails) correspondingQuestion
+                .getQuestionDetails()).getOtherEnabled();
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqResponseDetails.java]: You should consider to extract `((FeedbackMsqQuestionDetails) correspondingQuestion.getQuestionDetails())` to a variable.

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java (hunk 1 above)
Comment (RonakLakhotia) [reply src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: Yes, so should we use`-1`? or some other integer like `-999`?

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqResponseDetails.java [hunk 7]
```diff
@@ -85,7 +85,59 @@ public String getAnswerCsv(FeedbackQuestionDetails questionDetails) {
...
+            errors.add(Const.FeedbackQuestion.MSQ_ERROR_INVALID_OPTION);
+        }
+
+        // if selected answers are not a part of the Msq option list trigger this error
+        boolean isAnswersPartOfChoices = msqChoices.containsAll(answers);
+        if (!isAnswersPartOfChoices && !isNoneOfTheAboveOptionEnabled) {
+            errors.add(getAnswerString() + " " + Const.FeedbackQuestion.MSQ_ERROR_INVALID_OPTION);
+        }
+
+        // if other option is selected but not text is provided trigger this error
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackMsqResponseDetails.java]: no~~t~~ text

File: src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java (hunk 1 above)
Comment (wkurniawan07) [reply src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java]: I have suggested `-1` before.

**Comment by RonakLakhotia** [comment]
@wkurniawan07 updated changes.
